File: LintCode-Python/uglyNumberII.py
# Ugly number is a number that only have factors 2, 3 and 5.

# Design an algorithm to find the nth ugly number. The first 10 ugly numbers are
# 1, 2, 3, 4, 5, 6, 8, 9, 10, 12...

# Have you met this question in a real interview? Yes
#  Notice

# Note that 1 is typically treated as an ugly number.

# Example
# If n=9, return 10.

# Challenge
# O(n log n) or O(n) time.

def uglyNumberII(k):
    if not k:
        return None

    # Method 1
    # count = 1
    # result = 1
    # while count < k:
    #     result += 1
    #     num = result
    #     while num >= 2 and num % 2 == 0:
    #         num /= 2
    #     while num >= 3 and num % 3 == 0:
    #         num /= 3
    #     while num >= 5 and num % 5 == 0:
    #         num /= 5
    #     if num == 1:
    #         count += 1
    # return result

    # Method 2
    result = [1]
    curr = 2
    p2, p3, p5 = 0, 0, 0
    while len(result) < k:
        while 2 * result[p2] < curr:
            p2 += 1
        min2 = 2 * result[p2]
        while 3 * result[p3] < curr:
            p3 += 1
        min3 = 3 * result[p3]
        while 5 * result[p5] < curr:
            p5 += 1
        min5 = 5 * result[p5]
        next = min(min2, min3, min5)
        curr = next + 1
        result.append(next)
    return result[-1]

print uglyNumberII(9)

File: LeetCode_Java/src/UglyNumberII.java

/**
 * Write a program to find the n-th ugly number.
 * Ugly numbers are positive numbers whose prime factors only include 2, 3, 5. For example, 
 * 1, 2, 3, 4, 5, 6, 8, 9, 10, 12 is the sequence of the first 10 ugly numbers.
 * Note that 1 is typically treated as an ugly number.
 * Hint:
 * The naive approach is to call isUgly for every number until you reach the nth one. Most numbers 
 * are not ugly. Try to focus your effort on generating only the ugly ones.
 * An ugly number must be multiplied by either 2, 3, or 5 from a smaller ugly number.
 * The key is how to maintain the order of the ugly numbers. Try a similar approach of merging from 
 * three sorted lists: L1, L2, and L3.
 * Assume you have Uk, the kth ugly number. Then Uk+1 must be Min(L1 * 2, L2 * 3, L3 * 5).
 * 
 * Tags: DP, heap, Math
 * @author wendi
 *
 */
public class UglyNumberII {

	/**
	 * DP: use the definition of ugly number, the new ugly number always minimum of the old ugly 
	 * number *2, *3 and *5.
	 * Use int[] to save the old ugly number, and i2, i3 and i5 as pointers to record the position.
	 * @param int n
	 * @return int
	 * Time: O(n)
	 * Space: O(n)
	 */
	public int unlyNumberII(int n) {
		if (n <= 0) {
			return 0;
		}
		int[] ugly = new int[n];
		ugly[0] = 1;
		int i2 = 0;
		int i3 = 0;
		int i5 = 0;
		for (int i = 1; i < n; i++) {
			while (ugly[i2] * 2 <= ugly[i - 1]) {
				i2++;
			}
			while (ugly[i3] * 3 <= ugly[i - 1]) {
				i3++;
			}
			while (ugly[i5] * 5 <= ugly[i - 1]) {
				i5++;
			}
			ugly[i] = Math.min(Math.min(ugly[i2] * 2, ugly[i3] * 3), ugly[i5] * 5);
		}
		return ugly[n - 1];
	}
	
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		UglyNumberII result = new UglyNumberII();
		System.out.println(result.unlyNumberII(11));
	}

}
